File: Yixuan/16_01_17/last_position_of_target.java
public class Solution {
    /**
     * @param nums: An integer array sorted in ascending order
     * @param target: An integer
     * @return an integer
     * @link: http://www.lintcode.com/en/problem/last-position-of-target/
     * @author : Egbert Li
     */
    public int lastPosition(int[] nums, int target) {
      if (nums.length == 0) {
         return -1;
      }
      int start = 0;
      int end = nums.length - 1;
      while (start + 1 < end) {
         int mid = start + (end - start) / 2;
         if (target < nums[mid]) {
            end = mid;
         } else {
            start = mid;
         }
      }
      if (target == nums[end]) {
         return end;
      } else if (target == nums[start]) {
         return start;
      } else {
         return -1;
      }
    }
}

File: Boxuan/16_01_16/last_position_of_target.py
class Solution:
    # @param nums: The integer array
    # @param target: Target number to find
    # @return the first position of target in nums, position start from 0 
    def binarySearch(self, nums, target):
        
        # if not nums:
        #     return -1
        # for index, num in enumerate(nums[::-1]):
        #     if num == target:
        #         return index
        #     if num < target:
        #         return -1
        
        left, right = 0, len(nums)
        
        while left + 1 < right:
            mid = left + (right - left) / 2
            if nums[mid] > target:
                right = mid
            else:
                left = mid
                
        if nums[right] == target:
            return right
        if nums[left] == target:
            return left
            
        return -1
        